_posts/2020-06-30-sparklyr-1.3.0-released/sparklyr-1.3.0.Rmd

```diff
@@ -0,0 +1,175 @@
+
+
+## Higher-order Functions
+
+[Higher order functions](https://issues.apache.org/jira/browse/SPARK-19480) are built-in Spark SQL constructs that allow user-defined lambda expressions to be applied efficiently to complex data types such as arrays and structs. As a quick demo to see why higher-order functions are useful, let's say one day Scrooge Mcduck dived into his huge vault of money and found large quantities of pennies, nickels, dimes, and quarters. Having an impeccable taste in data structure, he decided to store the quantities and face values of everything into 2 Spark SQL array columns:
+
+```{bash echo=TRUE, eval=FALSE}
+$SPARK_HOME/bin/spark-sql
+spark-sql> CREATE TABLE `coins` (`quantities` ARRAY<INT>, `values` ARRAY<INT>) USING ORC;
+spark-sql> INSERT INTO `coins` VALUES (ARRAY(4000, 3000, 2000, 1000), ARRAY(1, 5, 10, 25));
+```
+
+thus declaring his net worth of 4k pennies, 3k nickels, 2k dimes, and 1k quarters. In sparklyr 1.3 or above, we can use a higher-order function named `ZIP_WITH` to help Scrooge Mcduck calculate the total value of each type of coin, and store the results in an array of length 4:
+
+```{bash echo=TRUE, eval=FALSE}
+spark-sql> SELECT ZIP_WITH(`quantities`, `values`, (quantity, value) -> quantity * value)
+         > AS `total_values` FROM `coins`;
+```
+```
+[4000,15000,20000,25000]
+```
+
+So the query result `[4000,15000,20000,25000]` indicates there are in total 40 dollars worth of pennies, 150 dollars worth of nickels, 200 dollars worth of dimes, and 250 dollars worth of quarters, as expected.
+
+With sparklyr 1.3, the same computation can be performed on a Spark data frame using `hof_zip_with()` with lambda expression above replaced by a [one-sided formula](https://stat.ethz.ch/R-manual/R-devel/library/base/html/tilde.html) of `.x` and `.y` in R, as shown below:
+
+```{r echo=TRUE, eval=FALSE}
+library(sparklyr)
+
+sc <- spark_connect(master = "local", version = "2.4")
+coins_tbl <- copy_to(
+  sc,
+  tibble::tibble(quantities = list(c(4000, 3000, 2000, 1000)), values = list(c(1, 5, 10, 25)))
+)
+
+result <- coins %>%
+  hof_zip_with(~ .x * .y, dest_col = total_values) %>%
+  dplyr::select(total_values) %>%
+  collect()
+
+result$total_values
+
+```
+```
+[[1]]
+[1]  4000 15000 20000 25000
+```
+
+Other higher-order functions supported by Spark SQL so far include `transform`, `filter`, `aggregate`, and `exists`, as documented in [here](https://spark.apache.org/docs/latest/api/sql/index.html), and similar to the example above, their counterparts (namely, `hof_transform`, `hof_filter`, `hof_aggregate`, and `hof_exists`) all exist in sparklyr 1.3, so that they can be integrated with other `dplyr` verbs in an idiomatic manner in R.
```

Comment:
```suggestion
Other higher-order functions supported by Spark SQL so far include `transform`, `filter`, `aggregate`, and `exists`, as documented in [here](https://spark.apache.org/docs/latest/api/sql/index.html), and similarly to the example above, their counterparts (namely, `hof_transform`, `hof_filter`, `hof_aggregate`, and `hof_exists`) all exist in sparklyr 1.3, so that they can be integrated with other `dplyr` verbs in an idiomatic manner in R.
```